objects/src/notes/recipient.rs

```diff
@@ -0,0 +1,111 @@
+use core::fmt::Debug;
+
+use super::{
+    ByteReader, ByteWriter, Deserializable, DeserializationError, Digest, Hasher, NoteInputs,
+    NoteScript, Serializable, Word,
+};
+
+/// Value that represents when a note can be consumed.
+///
+/// The recipient is not an account address, instead it is a value that describes when a note
+/// can be consumed. Because not all notes have predetermined consumer addresses, e.g. swap
+/// notes can be consumed by anyone, the recipient is defined as the code and its inputs, that
+/// when sucessfully executed results in the note's consumption.
+///
+/// Recipient is computed as:
+///
+/// > hash(hash(hash(serial_num, [0; 4]), script_hash), input_hash)
+///
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct NoteRecipient {
+    serial_num: Word,
+    script: NoteScript,
```

Comment:
This should probably be put into a `NoteNullifier`, but this PR was too big already.